File: dsa/educosys/Combinations.java
package educosys;

import java.util.List;
import java.util.ArrayList;

public class Combinations {

    public static void helper(int ind, int n, int k, List<Integer> curr, List<List<Integer>> res)
    {
        if (curr.size() == k)
        {
            res.add(new ArrayList<>(curr));
            return;
        }

        if (ind > n ) return;

        // include wala case
        curr.add(ind);
        helper(ind + 1, n, k, curr, res);
        curr.remove(curr.size() - 1);

        // exclude wala case
        helper(ind + 1, n, k, curr, res);
    }

    public static List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        helper(1, n, k, curr, res);
        return res;
    }

    public static void main(String[] args)
    {
        List<List<Integer>> res = combine(3, 2);
        for (List<Integer> list : res)
        {
            for (Integer val : list)
            {
                System.out.print(val + " ");
            }
            System.out.println();
        }
    }
    
}

File: dsa/educosys/combinations.cpp
#include <iostream>
#include <vector>
using namespace std;

void helper(int ind, int n, int k, vector<int>& curr, vector<vector<int>>& res)
{
    // base condition
    if (curr.size() == k)
    {
        res.push_back(curr);
        return;
    }

    if (ind > n) return;

    // include wala case
    curr.push_back(ind);
    helper(ind + 1, n, k, curr, res);
    curr.pop_back();

    // exclude wala case
    helper(ind + 1, n, k, curr, res);
}
vector<vector<int>> combine(int n, int k) {
    vector<int> curr;
    vector<vector<int>> res;
    helper(1, n, k, curr, res);
    return res;
}

int main(void) {
    vector<vector<int>> res = combine(3, 2);
    return 0;
}
